allow local NPC functions to be public or private

Helianthella opened the pull request:
This is my proposal to fix #2137:

* local functions can no longer be called as events, even if their label begins with `On`
* local functions may be explicitly exported by defining them as `public function ...`.
this allows other scripts to call them as if they were part of their own script
* exported local functions can be called from any script by doing `"npc name"::MyFunction();`

<br>

```c++
-	script	npc1	FAKE_NPC,{

	private function NonExported {
		return "[1] non-exported";
	}

	public function Exported {
		return "[1] exported";
	}

	// local call:
	debugmes(NonExported()); // => [1] non-exported
	debugmes(Exported()); // => [1] exported

	// call in another NPC:
	debugmes("npc2"::NonExported()); // ERROR: not exported
	debugmes("npc2"::Exported()); // => [2] exported
}


-	script	npc2	FAKE_NPC,{

	function NonExported {
		return "[2] non-exported";
	}

	public function Exported {
		return "[2] exported";
	}

	// local call:
	debugmes(NonExported()); // => [2] non-exported
	debugmes(Exported()); // => [2] exported

	// call in another NPC:
	debugmes("npc1"::NonExported()); // ERROR: not exported
	debugmes("npc1"::Exported()); // => [1] exported
}
```

```c++
// local call syntax (unchanged):
<function>({<arg>...})

// public call syntax:
"<npc name>"::<function>({<arg>...})

// also:
callfunctionofnpc("<npc name>", "<function>"{, <arg>...})
```

HerculesWSAPI commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/herculesws/hercules/2142)
<!-- Reviewable:end -->

mekolat commented on src/map/script.c:
```diff
@@ -1903,8 +1943,15 @@ static const char *parse_syntax(const char *p)
 
 			func_name = script->skip_space(p2);
+
+			if (*func_name == ':') {
```
to mark functions as exported I chose `:` because it is short and easy to remember, but we could use a keyword if this is preferred, such as:
* `extern function MyFunction { … }`
* `export function MyFunction { … }`
* `public function MyFunction { … }`

4144 commented on src/map/npc.h:
```diff
@@ -55,6 +55,7 @@ struct npc_timerevent_list {
 struct npc_label_list {
 	char name[NAME_LENGTH];
 	int pos;
+	uint8 flags; // is_extern, is_userfunc
```
probably better here use enum?

mekolat replied:
right, sorry

4144 commented on src/map/npc.c:
```diff
@@ -387,7 +387,7 @@ static int npc_event_export(struct npc_data *nd, int i)
 	lname = nd->u.scr.label_list[i].name;
 	pos = nd->u.scr.label_list[i].pos;
-	if ((lname[0] == 'O' || lname[0] == 'o') && (lname[1] == 'N' || lname[1] == 'n')) {
+	if ((nd->u.scr.label_list[i].flags & 0x1) && !(nd->u.scr.label_list[i].flags & 0x2)) {
```
0x1 and 0x2 probably better replace to enum values?

4144 commented:
and i think need add some tests to call local functions in CI

mekolat commented:
@4144 yes, sure, I will do this after I write the docs

mekolat commented on doc/script_commands.txt:
```diff
@@ -1963,18 +1967,31 @@ prontera,150,150,0	script	TestNPC	123,{
+*"<npc name>"::<function name>({<arg>...})
+
+In its first form, calls a previously defined local function. In its second
+form, calls a previously defined exported local function of another NPC.
```
`calls a previously defined exported local function of another NPC` … that's a mouthful,
but I don't really know how to explain it in simpler terms

mekolat commented on doc/script_commands.txt:
```diff
@@ -1904,57 +1904,61 @@ will result in error and termination of the script.
 ---------------------------------------
 
 *function <function name>;
-*<function name>{(<argument>, ...<argument>)};
```
something here bothers me: we're using curly braces to denote optional arguments, but in this case literal curly braces are required; how should this be written? escaped curly braces (`\{ <code> \}`) or a custom sequence like we do for `%TAB%`?

mekolat commented on doc/script_commands.txt:
```diff
@@ -1904,57 +1904,61 @@ will result in error and termination of the script.
 
 *function <function name>;
-*<function name>{(<argument>, ...<argument>)};
 *function <function name> {
```
something here bothers me: we're using curly braces to denote optional arguments, but in this case literal curly braces are required; how should this be written? escaped curly braces (`\{ <code> \}`) or a custom sequence like we do for `%TAB%`?

MishimaHaruna replied:
I think escaping them might cause more confusion, so I'd advise against that (or someone will really try to write it as `\{` in scripts).

I'd personally leave it as is (I don't believe it's ambiguous because it's not a function argument). Another option would be `%LCURLY%` / `%RCURLY%` but considering that I already find the `%TAB%` annoyingly hard to read...

mekolat commented:
@4144 I have added unit tests

mekolat commented:
moving this to the next milestone

Helianthella commented:
- Updated to use a `public` / `private` prefix instead of a colon
- Renamed the `foreigncall()` buildin that is used internally to `callfunctionofnpc()`
- Updated the documentation and also documented `callfunctionofnpc()`

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -848,57 +848,92 @@ static const char *parse_callfunc(const char *p, int require_paren, int is_custo
+				while( *name ) script->addb(*name ++);
+				script->addb(0);
+				arg = script->buildin[script->str_data[script->buildin_callfunc_ref].val];
+				if( *arg != '*' ) ++ arg;
```
Misplaced spaces and `++arg;` should be placed in new line.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -1240,6 +1275,13 @@ static const char *parse_simpleexpr(const char *p)
 		return script->parse_simpleexpr_number(p);
 	} else if(*p == '"') {
+		const char *p2 = p + 1;
+		while (*p2 != '"') ++p2;
```
`++p2;` should be placed in new line.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -1577,6 +1619,74 @@ static const char *parse_line(const char *p)
+		if (script->str_data[l].type == C_NOP) {
+			// register only, if the name was not used by something else
+			script->str_data[l].type = C_USERFUNC;
+		} else if(script->str_data[l].type == C_USERFUNC) {
```
Missing space after `if`.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -1577,6 +1619,74 @@ static const char *parse_line(const char *p)
+		// Close condition of if, for, while
+		p = script->parse_syntax_close(p2 + 1);
+		return p;
+	} else if(*p2 == '{') {
```
Missing space after `if`.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -1577,6 +1619,74 @@ static const char *parse_line(const char *p)
+			script->str_data[l].type = C_USERFUNC;
+			script->set_label(l, VECTOR_LENGTH(script->buf), p);
+
+			if (script->parse_options & SCRIPT_USE_LABEL_DB) {
```
Please use explicit comparison for non-boolean expressions.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -2674,11 +2746,14 @@ static struct script_code *parse_script(const char *src, const char *file, int l
 		// Special handling only label
 		tmpp=script->skip_space(script->skip_word(p));
-		if(*tmpp==':' && !(strncmp(p,"default:",8) == 0 && p + 7 == tmpp)) {
+		if(*tmpp==':' && !(strncmp(p,"default:",8) == 0 && p + 7 == tmpp) &&
```
Missing spaces and please break the line before the logical operator, not after.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -2674,11 +2746,14 @@ static struct script_code *parse_script(const char *src, const char *file, int l
 			script->set_label(i, VECTOR_LENGTH(script->buf), p);
-			if( script->parse_options&SCRIPT_USE_LABEL_DB )
-				script->label_add(i, VECTOR_LENGTH(script->buf));
+			if (script->parse_options & SCRIPT_USE_LABEL_DB) {
```
Please use explicit comparison for non-boolean expressions.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
+		if (strcmp(nd->u.scr.label_list[i].name, function_name) == 0) {
+			if (nd->u.scr.label_list[i].flags & LABEL_IS_EXTERN && nd->u.scr.label_list[i].flags & LABEL_IS_USERFUNC) {
+				pos = nd->u.scr.label_list[i].pos;
+			} else if (nd->u.scr.label_list[i].flags & LABEL_IS_USERFUNC) {
```
Please use explicit comparison for non-boolean expressions.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
+
+	for (i = 0; i < nd->u.scr.label_list_num; ++i) {
+		if (strcmp(nd->u.scr.label_list[i].name, function_name) == 0) {
+			if (nd->u.scr.label_list[i].flags & LABEL_IS_EXTERN && nd->u.scr.label_list[i].flags & LABEL_IS_USERFUNC) {
```
Please use explicit comparison for non-boolean expressions.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
+	st->stack->scope.vars = i64db_alloc(DB_OPT_RELEASE_DATA);
+	st->stack->scope.arrays = idb_alloc(DB_OPT_BASE);
+
+	if (!st->script->local.vars) {
```
Please use explicit comparison for non-boolean expressions.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
+	struct script_retinfo *ri = NULL;
+	struct reg_db *ref = NULL;
+	int pos = -1;
+	int i = 0;
```
Please fix scope of variables.

Kenpachi2k13 commented on src/map/script.h:
```diff
@@ -952,6 +954,7 @@ struct script_interface {
 	const char * (*parse_syntax_close) (const char *p);
 	const char * (*parse_syntax_close_sub) (const char *p, int *flag);
 	const char * (*parse_syntax) (const char *p);
+	const char * (*parse_syntax_function) (const char *p, bool is_public);
```
Asterisk should be adjacent to the field name.

Helianthella commented on src/map/script.c:
```diff
@@ -848,57 +848,92 @@ static const char *parse_callfunc(const char *p, int require_paren, int is_custo
+			arg = script->buildin[script->str_data[func].val];
+			if (script->str_data[func].deprecated)
+				DeprecationWarning(p);
+			if( !arg ) arg = &null_arg; // Use a dummy, null string
```
Most of the wrong formatting is actually just re-indented, unmodified legacy code but I don't mind refactoring it

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -848,121 +848,198 @@ static const char *parse_callfunc(const char *p, int require_paren, int is_custo
-		if( *p != ')' )
-			disp_error_message("parse_callfunc: expected ')' to close argument list",p);
+
+	if (arg && *arg && *arg != '?' && *arg != '*') {
```
`if (arg != NULL && *arg != 0 && ...`

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -1577,6 +1667,74 @@ static const char *parse_line(const char *p)
+			;  // already registered
+		} else {
+			disp_error_message("script:parse_syntax_function: function name is already in use", func_name);
+		}
```
To avoid that empty block you could change `== C_USERFUNC` to `!= C_USERFUNC` and show the error message in this block and remove the `else` block completely.
(Having an additional block just to insert a comment brings no benefit in my opinion.)

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -2668,25 +2788,32 @@ static struct script_code *parse_script(const char *src, const char *file, int l
-		if( *p == '\0' )
-			disp_error_message("unexpected end of script",p);
+	while (script->syntax.curly_count != 0 || *p != end) {
+		if(*p == '\0') {
```
Missing space.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -2668,25 +2788,32 @@ static struct script_code *parse_script(const char *src, const char *file, int l
-			i=script->add_word(p);
+		tmpp = script->skip_space(script->skip_word(p));
+
+		if (*tmpp==':' && !(strncmp(p, "default:", 8) == 0 && p + 7 == tmpp)
```
Missing spaces around `==`.

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -6397,6 +6524,89 @@ static BUILDIN(callfunc)
 }
+
+static BUILDIN(callfunctionofnpc) {
+	const char *function_name = script_getstr(st, 3);
```
Fix scope. (Move down to `int pos = -1;`)

Helianthella commented:
I'll do some commit squashing once the review process is over

Helianthella commented:
I added two config flags for more flexibility:
- `functions_private_by_default` defines whether functions are private or public when not explicitly set as public or private.
- `functions_as_events` allows functions to be called as events if made public. This allows for backward compatibility if someone was using this bug as a feature.

```c
-	script	@mycommand	FAKE_NPC,{
	public function OnInit {
		bindatcmd("mycommand", "@mycommand::OnCall");
		return;
	}

	private function doSomething {
		dispbottom(getarg(0));
		return;
	}

	public function OnCall {
		doSomething(.@atcmd_parameters$[0]);
		return;
	}
}
```

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -6397,6 +6530,89 @@ static BUILDIN(callfunc)
 	return true;
 }
+
+static BUILDIN(callfunctionofnpc) {
```
Please add a documentation. (Switching between code and script_commands.txt is a pain in the arse. :smile:)

Kenpachi2k13 commented on src/map/script.c:
```diff
@@ -1577,6 +1669,72 @@ static const char *parse_line(const char *p)
 	return p;
 }
 
+static const char *parse_syntax_function (const char *p, bool is_public)
```
Please add a documentation.

Kenpachi2k13 commented on src/map/map.h:
```diff
@@ -329,6 +329,11 @@ enum bl_type {
 
 enum npc_subtype { WARP, SHOP, SCRIPT, CASHSHOP, TOMB };
 
+enum script_label_flags {
```
Please add a documentation. (At least a small note about its purpose.)

Kenpachi2k13 commented:
Furthermore your commits need some squashing. :relaxed:

Helianthella commented:
@Kenpachi2k13 as noted earlier, I do intend on squashing once the review process is done.
It makes it easier for maintainers to review new changes when the branch is not force-pushed
